apm-sniffer/optional-plugins/optional-spring-plugins/optional-spring-cloud/gateway-2.0.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/cloud/gateway/v20x/HttpClientRequestInterceptor.java

```diff
@@ -80,7 +84,7 @@ public Publisher<Void> apply(final HttpClientRequest httpClientRequest) {
             }
         };
 
-        objInst.setSkyWalkingDynamicField(new EnhanceCacheObject(span, abstractSpan));
+        ENHANCE_CONTEXT.set(new EnhanceCacheObject(span, abstractSpan));
```

Comment:
> The HttpClient instance is a singleton, and the setting and retrieval of EnhanceCacheObject are not thread-safe. Therefore, the EnhanceCacheObject obtained in the afterMethod may not necessarily be set by the same thread in the beforeMethod.

In fact, this description is not correct, no need for threadlocal here

Reply:
In the beforeMethod method, the name of the thread is passed to EnhanceCacheObject, but in the afterMethod method, when it is retrieved and compared to the current thread, they are not consistent, and the URL is also not consistent.

Reply:
step:
1. thread1 -> beforeMethod -> set EnhanceCacheObject1
2. thread2 -> beforeMethod -> set EnhanceCacheObject2
3. thread1 -> afterMethod -> get EnhanceCacheObject2
4. thread2 -> afterMethod -> get EnhanceCacheObject2

Reply:
Reproduction steps：
In a high-concurrency scenario, the allArguments[1] is stored in EnhanceCacheObject in the beforeMethod method, but in the afterMethod method, the allArguments[1] of the method parameter is different from the allArguments[1] in EnhanceCacheObject.
Similarly, the thread name can also be reproduced.

Reply:
dynamicField is instance level, not thread safe.
If you just need context between before and after, please use v2 API, which has MethodInvokeContext for this.

Reply:
https://skywalking.apache.org/docs/skywalking-java/next/en/setup/service-agent/java-agent/java-plugin-development-guide/#v2-apis-1

Reply:
> dynamicField is instance level, not thread safe. If you just need context between before and after, please use v2 API, which has MethodInvokeContext for this.

I will make the changes later.

Reply:
Okay, I misunderstood before，I misunderstood it as an enhanced instance of passing in parameters.

Reply:
That was why we introduced v2.
I believe it would make this kind of scenarios easier.

Reply:
> That was why we introduced v2. I believe it would make this kind of scenarios easier.

I have a question. I just found out in the document that there are both v1 and v2 versions of ClassEnhancePluginDefine. Do I have to use ClassEnhancePluginDefineV2 to define InstanceMethodsAroundInterceptorV2?

Reply:
> > That was why we introduced v2. I believe it would make this kind of scenarios easier.
> 
> I have a question. I just found out in the document that there are both v1 and v2 versions of ClassEnhancePluginDefine. Do I have to use ClassEnhancePluginDefineV2 to define InstanceMethodsAroundInterceptorV2?

```java
Caused by: java.lang.ClassCastException: org.apache.skywalking.apm.plugin.spring.cloud.gateway.v20x.HttpClientRequestInterceptor cannot be cast to org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor
```

From the exception message, it seems that you need to modify the definition.

Reply:
Yes. They are using different kernel codes. You need to align them.

Reply:
> Yes. They are using different kernel codes. You need to align them.

I have already made the modification, and it's running fine on my project.